api/src/services/users.js

```diff
@@ -565,16 +619,70 @@ module.exports = {
-      .then(() => tokenLogin.manageTokenLogin(data, appUrl, response))
-      .then(() => response);
+    await createUserEntities(data, response, appUrl);
+    return response;
+  },
+
+  /**
+   * Take the request data and create valid users, user-settings and contacts
+   * objects. Returns the response body in the callback.
+   * @param {Object|Object[]} users
+   * @param {string} users[].username
+   * @param {string=} users[].phone Not required if the password is provided.
+   * @param {string=} users[].password Not required if the phone number is provided.
+   * @param {string[]} users[].roles
+   * @param {(Object|string)=} users[].place Can either be a place object or an existing place id.
+   * @param {(Object|string)=} users[].contact Can either be a contact object or an existing place id.
+   * @param {string=} users[].type Deprecated. Used to infer user's roles
+   * @param {string} appUrl   request protocol://hostname
+   */
+  async createUsers(users, appUrl) {
+    if (!Array.isArray(users)) {
+      return module.exports.createUser(users, appUrl);
+    }
+
+    const { missingFieldsFailingIndexes, tokenLoginFailingIndexes, passwordFailingIndexes } = validateUserFields(users);
+    if (missingFieldsFailingIndexes.length > 0) {
+      let errorMessage = 'Missing required fields:\n';
+      for (const { fields, index } of missingFieldsFailingIndexes) {
+        errorMessage += `\nMissing fields ${fields.join(', ')} for user at index ${index}`;
+      }
+      return Promise.reject(error400(errorMessage, { failingIndexes: missingFieldsFailingIndexes }));
+    }
+
+    if (tokenLoginFailingIndexes.length > 0) {
+      let errorMessage = 'Token login errors:\n';
+      for (const { tokenLoginError, index } of tokenLoginFailingIndexes) {
+        errorMessage += `\nError ${tokenLoginError.msg} for user at index ${index}`;
+      }
+      return Promise.reject(error400(errorMessage, { failingIndexes: tokenLoginFailingIndexes }));
+    }
+
+    if (passwordFailingIndexes.length > 0) {
+      let errorMessage = 'Password errors:\n';
+      for (const { passwordError, index } of passwordFailingIndexes) {
+        errorMessage += `\nError ${passwordError.message.message} for user at index ${index}`;
+      }
+      return Promise.reject(error400(errorMessage, { failingIndexes: passwordFailingIndexes }));
+    }
+
+    const response = Array(users.length).fill(null).map(() => ({}));
+    // create all valid users even if some are failing
+    const promises = await allPromisesSettled(
+      users.map(async (user, index) => await createUserEntities(user, response[index], appUrl)),
+    );
+    const hasFailures = promises.some(promise => promise.status === 'rejected');
+    if (hasFailures) {
+      promises.forEach((promise, index) => {
+        if (promise.status === 'rejected') {
+          response[index].error = promise.reason.message;
+        }
+      });
+    }
+
+    return response;
```

Comment:
I we change `createUserEntities` to create and return the response object., this whole section can be rewritten like: 
```suggestion
    const promises = await allPromisesSettled(users.map((user) => createUserEntities(user, appUrl)));
    return promises.map(promise => {
      return promise.status === 'fulfilled' ? promise.value : { error: promise.reason.message };
    });
```